File: alexandria/ui/src/io/intino/alexandria/ui/displays/components/GroupBox.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.notifiers.GroupBoxNotifier;
import io.intino.alexandria.ui.model.Datasource;

import java.util.Arrays;
import java.util.List;

public class GroupBox<DN extends GroupBoxNotifier, B extends Box> extends AbstractGroupBox<B> {
    private Collection collection;
    private List<String> selection;

    public GroupBox(B box) {
        super(box);
    }

    public GroupBox<DN, B> bindTo(Collection collection) {
        this.collection = collection;
        this.collection.onRefresh((event) -> refresh(event.items()));
        return this;
    }

    public void select(String[] groups) {
        this.selection = Arrays.asList(groups);
        collection.filter(name(), selection);
    }

    public void refresh(List<Object> items) {
        super.refresh();
        Datasource source = collection.source();
        List groups = source.groups(name(), items);
        notifier.refreshGroups(groups);
    }
}

File: framework/ui-elements/src/displays/components/GroupBox.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractGroupBox from "../../../gen/displays/components/AbstractGroupBox";
import GroupBoxNotifier from "../../../gen/displays/notifiers/GroupBoxNotifier";
import GroupBoxRequester from "../../../gen/displays/requesters/GroupBoxRequester";
import { List, ListItem, ListItemText, ListItemSecondaryAction, Checkbox, Typography } from "@material-ui/core";
import 'alexandria-ui-elements/res/styles/layout.css';

const styles = theme => ({
	label : {
		padding: "10px 10px 0"
	},
	checkbox : {
		padding: "0 7px"
	},
	count : {
		color: theme.palette.grey.primary
	},
	group : {
		paddingLeft : "0"
	}
});

class GroupBox extends AbstractGroupBox {
	state = {
		groups : [],
		selection: []
	};

	constructor(props) {
		super(props);
		this.notifier = new GroupBoxNotifier(this);
		this.requester = new GroupBoxRequester(this);
	};

	render() {
		return (
			<div style={this.style()}>
				{this.props.label != null ? <Typography className={this.props.classes.label} variant="subtitle1">{this.props.label}</Typography> : undefined}
				<List>{this.state.groups.map((group, i) => this.renderGroup(group, i))}</List>
			</div>
		);
	}

	renderGroup = (group, index) => {
		const { classes } = this.props;
		return (
			<ListItem key={index} className={classes.group} role={undefined} dense button onClick={this.handleToggle(group)}>
				<Checkbox className={classes.checkbox} checked={this.state.selection.indexOf(group.label) !== -1} tabIndex={-1} disableRipple/>
				<ListItemText>{group.label}</ListItemText>
				<ListItemSecondaryAction className={classes.count}>{group.count}</ListItemSecondaryAction>
			</ListItem>
		);
	};

	handleToggle = group => () => {
		const { selection } = this.state;
		const currentIndex = selection.indexOf(group.label);
		const newSelection = [...selection];

		if (currentIndex === -1) newSelection.push(group.label);
		else newSelection.splice(currentIndex, 1);

		this.requester.select(newSelection);
		this.setState({ selection: newSelection });
	};

	refreshGroups = (groups) => {
		this.setState({ groups });
	};
}

export default withStyles(styles, { withTheme: true })(GroupBox);